constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/Windows.java

```diff
@@ -1,267 +1,327 @@
+    result.set(interval, value);
     return result;
   }
 
-
-  public void subtract(final Window window) {
-    this.windows.subtract(window);
-  }
-
-  public void subtractAll(final Windows other) {
-    this.windows.subtractAll(other.windows);
-  }
-
-  public void subtract(final long start, final long end, final Duration unit) {
-    this.subtract(Window.between(start, end, unit));
-  }
-
-  public void subtractPoint(final long quantity, final Duration unit) {
-    this.subtract(Window.at(quantity, unit));
-  }
-
-  public static Windows minus(final Windows left, final Windows right) {
-    final var result = new Windows(left);
-    result.subtractAll(right);
+  public static Windows definedEverywhere(final List<Interval> intervals, final boolean value) {
+    final var result = new Windows(Interval.FOREVER, !value);
+    result.setAll(intervals, value);
     return result;
   }
 
-  public void intersectWith(final Window window) {
-    this.windows.intersectWith(window);
-  }
-
-  public void intersectWith(final Windows other) {
-    this.windows.intersectWithAll(other.windows);
-  }
-
-  public void intersectWith(final long start, final long end, final Duration unit) {
-    this.intersectWith(Window.between(start, end, unit));
+  public void setAllTrue(final List<Interval> intervals) {
+    for (final var interval: intervals) {
+      this.set(interval, true);
+    }
   }
 
-  public Optional<Duration> minTimePoint(){
-    if(!isEmpty()) {
-      return Optional.of(this.windows.ascendingOrder().iterator().next().start);
+  public void setTrue(final Interval... intervals) {
+    this.setAllTrue(Arrays.stream(intervals).toList());
+  }
+
+  public Windows and(final Windows other) {
+    return new Windows(IntervalMap.map2(
+        this, other,
+        (l, r) -> {
+          if (l.isPresent() && r.isPresent()) {
+            return Optional.of(l.get() && r.get());
+          } else if (l.isPresent()) {
+            return l.get() ? Optional.empty() : Optional.of(Boolean.FALSE);
+          } else if (r.isPresent()) {
+            return r.get() ? Optional.empty() : Optional.of(Boolean.FALSE);
+          } else {
+            return Optional.empty();
+          }
+        }
+    ));
+  }
+
+  public Windows and(final Interval other) {
+    return this.and(definedEverywhere(other, true));
+  }
+
+  public Windows or(final Windows other) {
+    return new Windows(IntervalMap.map2(
+        this, other,
+        (l, r) -> {
+          if (l.isPresent() && r.isPresent()) {
+            return Optional.of(l.get() || r.get());
+          } else if (l.isPresent()) {
+            return l.get() ? Optional.of(true) : Optional.empty();
+          } else if (r.isPresent()) {
+            return r.get() ? Optional.of(true) : Optional.empty();
+          } else {
+            return Optional.empty();
+          }
+        }
+    ));
+  }
+
+  public Windows or(final Interval other) {
+    return this.or(definedEverywhere(other, true));
+  }
```

Comment:
Giving this [the same treatment](https://github.com/NASA-AMMOS/aerie/pull/295/files#r952012701) as `and(Interval)`, `or` is essentially an overlay.

```java
  public Windows or(final Interval interval) {
    return new Windows(IntervalMap.map2(
        this, new Windows(interval, true),
        (l, r) -> (r.isPresent()) ? r : l));
  }
```

Generalized:
```java
  public IntervalMap<V> withOverlay(final IntervalMap<V> overlay {
    return IntervalMap.map2(this, overlay, (l, r) -> (r.isPresent()) ? r : l));
  }
```

So `and(Interval)` is a selection operation, and `or(Interval)` is an overlay operation.

Reply:
that one's just `set`

Reply:
Cool! The generalized form is just `setAll`, in particular.

So we've already got two very nice immutable replacements for some of the mutable forms.

* `bound` is replaced by `mask` / `select` (pick your favorite, I like `select`)
* `setAll` is replaced by `withOverlay`